File: BOJ/29_유니온파인드/1717_집합의표현.py
import sys

read = sys.stdin.readline
sys.setrecursionlimit(int(1e9))
def find(x) :
    if par[x] == x : return par[x]
    else : 
        par[x] = find(par[x])
        return par[x]

def merge(x,y) :
    p_x = find(x)
    p_y = find(y)

    par[p_x] = p_y



N,M = map(int,read().split())
par = [i for i in range(N+1)]
for _ in range(M) :
    cmd,x,y = map(int,read().split())

    if cmd == 0 : 
        merge(x,y)

    else :
        if find(x) == find(y) :
            print("YES")
        else : 
            print("NO")

File: BOJ/29_유니온파인드/1717_집합의표현.cpp
#include <iostream>

#define MAXN 1000000

using namespace std;

int N,M,cmd,x,y;
int par[MAXN+1];
int find(int x) {
    if(par[x] == x) return par[x];
    else return par[x] = find(par[x]);
}

void merge(int x, int y) {
    int p_x = find(x);
    int p_y = find(y);

    par[p_x] = p_y;
}

int main() {

    scanf("%d %d",&N,&M);
    for(int i=0; i<=N; i++) par[i] = i;
    for(int i=0; i<M; i++) {
        scanf("%d %d %d",&cmd,&x,&y);

        if(cmd == 0) merge(x,y);
        else {
            if(find(x) == find(y)) printf("YES\n");
            else printf("NO\n");
        }
    }
    return 0;
}